modules/ting_subsearch/ting_subsearch_suggestions/ting_subsearch_suggestions.module

```diff
@@ -0,0 +1,297 @@
+        ];
+        $query = ting_subsearch_get_query_string($additional);
+        global $base_url;
+        $url = $base_url . '/search/ting/' . $suggestion . $query;
+
+        header('Location: ' . $url, TRUE, 301);
+        drupal_exit($url);
+      }
+    }
+  }
+
+  $output = '';
+
+  // If we're already redirecting from a zero-hit search avoid trying to find
+  // suggestions for our own suggestion. Display a zero hit suggestions message.
+  if (!empty($_GET['original-search'])) {
+    $output = theme('ting_subsearch_message', [
+      'message' => t('Showing results for <strong>"@suggestion"</strong>. The search for <strong>"@original_search"</strong> returned 0 hits.', [
+        '@suggestion' => $keys,
+        '@original_search' => $_GET['original-search'],
+      ]),
+      'type' => 'suggestions',
+      'suggestion' => $keys,
+      'suggestion_num_results' => $original_search_num_results,
+      'original_search' => $_GET['original-search'],
+      'original_search_num_results' => 0,
+    ]);
+  }
+  // Otherwise trigger non-zero-hit seconday search if it's below limit.
+  // For performance reasons we handel this with async AJAX request, since it's
+  // not as critical as a zero-hit search.
+  elseif ($original_search_num_results <= variable_get("ting_subsearch_suggestions_trigger_limit", 5)) {
+    drupal_add_js([
+      'tingSubsearchSuggestions' => [
+        'originalSearch' => $keys,
+        'originalSearchNumResults' => $original_search_num_results,
+        'query' => ting_subsearch_get_query_string(),
+      ],
+    ], 'setting');
+
+    drupal_add_js(drupal_get_path('module', 'ting_subsearch_suggestions') . '/js/subsearch_suggestions.js', [
+      'type' => 'file',
+      'scope' => 'footer',
+    ]);
+
+    // Output a placeholder to be replaced with the actual content returned
+    // from AJAX callback.
+    $output = '<div id="ting-subsearch-suggestions-placeholder"></div>';
+  }
+
+  return $output;
+}
+
+/**
+ * Find the best suitable suggestion for the given original search.
+ *
+ * @param string $original_search
+ *   The original search that you want to find suggestion for.
+ *
+ * @return string
+ *   The suggestion or FALSE if it was not possible to find a suggestion within
+ *   the configured constraints.
+ *
+ * @throws Exception
+ */
+function ting_subsearch_suggestions_suggest_keys($original_search) {
+  if (!ding_provider_implements('search', 'autocomplete_suggestions')) {
+    return FALSE;
+  }
+
+  $original_search = ting_subsearch_normalize_keys($original_search);
+
+  if (strlen($original_search) > TING_SUBSEARCH_SUGGESTIONS_LEVENSHTEIN_LIMIT) {
+    return FALSE;
+  }
+
+  // Build expected query object for provider hook.
+  $query = [];
+  $query['q'] = $original_search;
+  $query['type'] = 'all';
+  $suggestions = ding_provider_invoke('search', 'autocomplete_suggestions', $query);
+
+  $result = array_reduce($suggestions, function($carry, $suggestion) use ($original_search) {
```

Comment:
Sorry I completely forgot comments here!

We can't use the suggestions directly. For example for author suggestions, it will often return something like emelie schepp (f. 1979). [Exampe](https://ortograf.dbc.dk/ortograf/suggest?type=all&q=emilie%20schepp)

So this screws up the levenshtein distance comparison.

So we build up a candidate word for word until it has reached an acceptable levenshtein distance. Continue with next suggestion if it gets too long or we run out of words.

Will add some better comments.

Reply:
I've added some more comments and refactores the code. Hopefully it's more understandable now :)